server/handlers/project_command_output_handler.go

```diff
@@ -172,40 +189,50 @@ func (p *AsyncProjectCommandOutputHandler) writeLogLine(pull string, line string
 	}
 
 	p.projectOutputBuffersLock.Lock()
-	if p.projectOutputBuffers[pull] == nil {
-		p.projectOutputBuffers[pull] = []string{}
+	if p.projectOutputBuffers[jobID] == nil {
+		p.projectOutputBuffers[jobID] = []string{}
 	}
-	p.projectOutputBuffers[pull] = append(p.projectOutputBuffers[pull], line)
+	p.projectOutputBuffers[jobID] = append(p.projectOutputBuffers[jobID], line)
 	p.projectOutputBuffersLock.Unlock()
 }
 
 //Remove channel, so client no longer receives Terraform output
-func (p *AsyncProjectCommandOutputHandler) Deregister(pull string, ch chan string) {
-	p.logger.Debug("Removing channel for %s", pull)
+func (p *AsyncProjectCommandOutputHandler) Deregister(jobID string, ch chan string) {
+	p.logger.Debug("Removing channel for %s", jobID)
 	p.receiverBuffersLock.Lock()
-	delete(p.receiverBuffers[pull], ch)
+	delete(p.receiverBuffers[jobID], ch)
 	p.receiverBuffersLock.Unlock()
 }
 
-func (p *AsyncProjectCommandOutputHandler) GetReceiverBufferForPull(pull string) map[chan string]bool {
-	return p.receiverBuffers[pull]
+func (p *AsyncProjectCommandOutputHandler) GetReceiverBufferForPull(jobID string) map[chan string]bool {
+	return p.receiverBuffers[jobID]
 }
 
-func (p *AsyncProjectCommandOutputHandler) GetProjectOutputBuffer(pull string) []string {
-	return p.projectOutputBuffers[pull]
+func (p *AsyncProjectCommandOutputHandler) GetProjectOutputBuffer(jobID string) []string {
+	return p.projectOutputBuffers[jobID]
 }
 
-func (p *AsyncProjectCommandOutputHandler) CleanUp(pull string) {
-	p.projectOutputBuffersLock.Lock()
-	delete(p.projectOutputBuffers, pull)
-	p.projectOutputBuffersLock.Unlock()
+func (p *AsyncProjectCommandOutputHandler) GetJobIdMapForPullContext(pullContext models.PullContext) map[string]bool {
+	return p.pullToJobMapping[pullContext]
+}
 
-	// Only delete the pull record from receiver buffers.
-	// WS channel will be closed when the user closes the browser tab
-	// in closeHanlder().
-	p.receiverBuffersLock.Lock()
-	delete(p.receiverBuffers, pull)
-	p.receiverBuffersLock.Unlock()
+func (p *AsyncProjectCommandOutputHandler) CleanUp(pullContext models.PullContext) {
+	if jobIdMap, ok := p.pullToJobMapping[pullContext]; ok {
```

Comment:
we still need some sort of mutex here whether it's `sync.Map` or mutexes.

Reply:
Decided to use sync.Map!